### PR TITLE
Refactor FPD Merge Functions

### PR DESCRIPTION
**Pull request by SyntaxNode**
Replaces the FPD merge methods with a smarter implementation.

With the goal of parsing the FPD overlay json only once, the exisitng algorithm cloned the entire app / site / user object before unmarshalling the overlay. A clone is necessary because we share object references between bidder requests to save on allocations, so we need to make a clone to a pointer field before we make a bidder specific change.

Instead of cloning the entire object up front, we can save on allocations and processing time in most cases by tying into the unmarshal process and making a clone as needed. This new approach (now possible with the move to json-iter) also builds ext merging into the unmarshal process and eliminates the need for the extMerger helper.

Benchmarks:
```
Smallest:
App + ID, ID Overwritten
BenchmarkAppCopy-12    	          976220	        1073 ns/op	     872 B/op	      14 allocs/op
BenchmarkAppMergeClone-12      	 1706986	       700.8 ns/op	     560 B/op	      11 allocs/op

Small:
App - ID + Publisher, ID + Publisher Overwritten
BenchmarkAppCopy-12          	 1883467	       664.5 ns/op	     872 B/op	       8 allocs/op
BenchmarkAppMergeClone-12    	 2491872	       517.9 ns/op	     576 B/op	       6 allocs/op

Medium / Average Case
App - Some Set, ID + Publisher Overwritten
BenchmarkAppCopy-12         	   95799	     12359 ns/op	    8708 B/op	     164 allocs/op
BenchmarkAppMergeClone-12         449222	      2645 ns/op	    1856 B/op	      44 allocs/op

Medium - All Fields Set On Request
App - All Set, ID + Publisher Overwritten
BenchmarkAppCopy-12         	   58051	     21394 ns/op	   14798 B/op	     280 allocs/op
BenchmarkAppMergeClone-12         928129	      1257 ns/op	    1760 B/op	      34 allocs/op

Largest / Worst:
App - All Set, All Overwritten
BenchmarkAppCopy-12         	   34359	     33525 ns/op	   17426 B/op	     405 allocs/op
BenchmarkAppMergeClone-12          33030	     37734 ns/op	   19092 B/op	     435 allocs/op
```

This new algorithm uses less cpu time and allocates less in almost all cases, except for the worst case. I wouldn't expect the worst case to ever occur, but sharing the benchmark for full transparency. The new algorithm really shines for the medium / average cases.

### CODE REVIEW

**Review by VeronikaSolovei9 — DISMISSED** [review]

File: util/jsonutil/merge.go [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+	mapType      *reflect2.UnsafeMapType
+	valueDecoder jsoniter.ValDecoder
+}
+
+func (d *mapCloneDecoder) Decode(ptr unsafe.Pointer, iter *jsoniter.Iterator) {
+	// don't clone if field is being set to nil. checking for nil "consumes" the null
+	// token, so must be handled in this decoder.
+	if iter.ReadNil() {
+		*(*unsafe.Pointer)(ptr) = nil
+		d.mapType.UnsafeSet(ptr, d.mapType.UnsafeNew())
```
Comment (guscarreon) [inline util/jsonutil/merge.go:128]: Is this call initializing `ptr` to a new empty map? If so, slices and pointers in lines 80 and 104 are simply set to `nil`.

**Review by guscarreon — COMMENTED** [review]
I didn't find the benchmark in the contributed code. Should we include them?

File: util/jsonutil/merge.go (hunk 1 above)
Comment (SyntaxNode) [reply util/jsonutil/merge.go:128]: I do not fully understand why setting the pointer to `nil` isn't sufficient for maps. This is copy I copied/pasted from json-iter, see: https://github.com/json-iterator/go/blob/71ac16282d122fdd1e3a6d3e7f79b79b4cc3b50e/reflect_map.go#L158.

File: util/jsonutil/merge_test.go [hunk 2]
```diff
@@ -0,0 +1,443 @@
...
+	})
+
+	t.Run("invalid-null", func(t *testing.T) {
+		var (
+			banner = &openrtb2.Banner{ID: "1"}
+			imp    = &openrtb2.Imp{Banner: banner}
+		)
+
+		err := MergeClone(imp, []byte(`{"banner":nul}`))
+		require.EqualError(t, err, "cannot unmarshal openrtb2.Imp.Banner: expect ull")
```
[the remark this answers is not in the document]
Comment (SyntaxNode) [reply util/jsonutil/merge_test.go:82]: This is not a bug in this PR. That is the error message from json-iter. I'll add a comment to clarify.

File: util/jsonutil/merge_test.go [hunk 3]
```diff
@@ -0,0 +1,443 @@
...
+
+	t.Run("invalid-malformed", func(t *testing.T) {
+		var (
+			banner = &openrtb2.Banner{ID: "1"}
+			imp    = &openrtb2.Imp{Banner: banner}
+		)
+
+		err := MergeClone(imp, []byte(`{"banner":malformed}`))
+		require.EqualError(t, err, "cannot unmarshal openrtb2.Imp.Banner: expect { or n, but found m")
+	})
```
[the remark this answers is not in the document]
Comment (SyntaxNode) [reply util/jsonutil/merge_test.go:97]: No. It is proper to parse an empty json array to an empty slice. There is sometimes a difference between a null and empty array.

**Review by guscarreon — APPROVED** [review]
LGTM

**Review by VeronikaSolovei9 — APPROVED** [review]

**Pull request merged**